### ISSUE TITLE
Getting 'badrequest' error deploying ARM

### ISSUE THREAD

**Issue by onionhammer** [Azure/azure-functions-host#6568]
When I try to deploy an ARM template with an output:

```json
  "outputs": {
    "armkey": {
      "type": "string",
      // "value": "[listKeys(resourceId('Microsoft.Web/sites', variables('workerApp')), '2019-08-01')]"
      // "value": "[listkeys(concat(resourceId('Microsoft.Web/sites', variables('workerApp')), '/host/default/'),'2016-08-01').functionKeys.default]"
      "value": "[listkeys(concat(resourceId('Microsoft.Web/sites', variables('workerApp')), '/host/default/'),'2016-08-01').functionKeys.default]"
    }
  }
```

I'm receiving:

```json
{
    "Code": "BadRequest",
    "Message": "Encountered an error (ServiceUnavailable) from host runtime.",
    "Target": null,
    "Details": [
        {
            "Message": "Encountered an error (ServiceUnavailable) from host runtime."
        },
        {
            "Code": "BadRequest"
        },
        {
            "ErrorEntity": {
                "Code": "BadRequest",
                "Message": "Encountered an error (ServiceUnavailable) from host runtime."
            }
        }
    ],
    "Innererror": null
}
```

**Comment by balag0** [comment]
Can you share the name of function app, time stamp & region?

Also please take a look at https://github.com/Azure/azure-functions-host/issues/5722

**Comment by onionhammer** [comment]
Name of the app is next-c2c-worker, central US, 8/27/2020, 5:48:11 PM central time. I think the issue is whenever I re-run the deploy it seems to clear out the function app until I redeploy..

**Comment by onionhammer** [comment]
I think the issue here is that ARM deploys seem to randomly clear out the entire function, and then the error returned by the management API to listkey() is the above if the function app is empty.. the ARM template will never work the first time for a new resource group, right?

**Comment by onionhammer** [comment]
When I deploy via ARM it removes the 'WEBSITE_RUN_FROM_PACKAGE' value, even though it is not present in the ARM template; is ARM able to somehow merge that value in?

**Comment by balag0** [comment]
I think any setting not explicitly mentioned would get updated to its default value which in this case would remove it.  Is it possible for you to include the setting WEBSITE_RUN_FROM_PACKAGE for each template deployment?

https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/deployment-modes#incremental-mode

 Note

When redeploying an existing resource in incremental mode, all properties are reapplied. The properties aren't incrementally added. A common misunderstanding is to think properties that aren't specified in the template are left unchanged. If you don't specify certain properties, Resource Manager interprets the deployment as overwriting those values. Properties that aren't included in the template are reset to the default values. Specify all non-default values for the resource, not just the ones you're updating. The resource definition in the template always contains the final state of the resource. It can't represent a partial update to an existing resource.

**Comment by onionhammer** [comment]
@balag0 no, since WEBSITE_RUN_FROM_PACKAGE is set to some URL which is generated by the deployment.. so basically updating ARM templates from source control is not supported, which IMO kind of defeats the purpose of repeatable deployments via ARM.

**Comment by Crossbow78** [comment]
Note that we get the same error with the Az CLI tools, which introduced `functionapp keys` management in a recent version. This both happens locally and in the Azure Portal cloud shell.
Also the Azure portal itself does not seem to list any keys in the UI anymore.

When trying to 'list' the available keys in our functionapp, the error response in the Az CLI is:
```
urllib3.connectionpool : Starting new HTTPS connection (1): management.azure.com:443
urllib3.connectionpool : https://management.azure.com:443 "POST /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/xxxx-xxx-xxx/providers/Microsoft.Web/sites/xxxx-xxx-xxx/host/default/listkeys?api-version=2019-08-01 HTTP/1.1" 400 349
msrest.http_logger : Response status: 400
msrest.http_logger : Response headers:
msrest.http_logger :     'Cache-Control': 'no-cache'
msrest.http_logger :     'Pragma': 'no-cache'
msrest.http_logger :     'Content-Length': '349'
msrest.http_logger :     'Content-Type': 'application/json; charset=utf-8'
msrest.http_logger :     'Expires': '-1'
msrest.http_logger :     'Strict-Transport-Security': 'max-age=31536000; includeSubDomains'
msrest.http_logger :     'Server': 'Microsoft-IIS/10.0'
msrest.http_logger :     'X-AspNet-Version': '4.0.30319'
msrest.http_logger :     'X-Powered-By': 'ASP.NET'
msrest.http_logger :     'x-ms-ratelimit-remaining-subscription-writes': '1199'
msrest.http_logger :     'x-ms-request-id': '0405495c-08c5-44fb-ad81-cce6caa072d5'
msrest.http_logger :     'x-ms-correlation-request-id': '0405495c-08c5-44fb-ad81-cce6caa072d5'
msrest.http_logger :     'x-ms-routing-request-id': 'WESTEUROPE:20200903T094546Z:0405495c-08c5-44fb-ad81-cce6caa072d5'
msrest.http_logger :     'X-Content-Type-Options': 'nosniff'
msrest.http_logger :     'Date': 'Thu, 03 Sep 2020 09:45:45 GMT'
msrest.http_logger : Response content:
msrest.http_logger : {"Code":"BadRequest","Message":"Encountered an error (InternalServerError) from host runtime.","Target":null,"Details":[{"Message":"Encountered an error (InternalServerError) from host runtime."},{"Code":"BadRequest"},{"ErrorEntity":{"Code":"BadRequest","Message":"Encountered an error (InternalServerError) from host runtime."}}],"Innererror":null}
msrest.exceptions : Operation returned an invalid status code 'Bad Request'
cli.azure.cli.core.util : Operation returned an invalid status code 'Bad Request'
Operation returned an invalid status code 'Bad Request'
```

**Edit (24-09):** It turned out that our function app was unable to reach its storage account due to a misconfigured VNet integration, which resulted in the internal server error when retrieving the app keys.

**Comment by alona-shevliakova** [comment]
hey, is somebody found any solution for this?

**Comment by v-anvari** [comment]
Hi @onionhammer , Apologies for the delayed response. Were you able find a solution? do let us know if a follow up is required

**Comment by onionhammer** [comment]
@v-anvari no there is no solution; if you re-run an ARM deployment against an azure function it will effectively wipe out whats currently deployed out there and you will have to deploy it again, unlike other apps running in app service

**Comment by v-anvari** [comment]
Hi @onionhammer , as Bala earlier explained in his comment that, If you don't specify certain properties, Resource Manager interprets the deployment as overwriting those values. Properties that aren't included in the template are reset to the default values. So, this is as per design. Let us know if you have any further queries

**Comment by onionhammer** [comment]
@v-anvari Right, I'm saying the design is flawed. Currently if you re-run an ARM deployment of an Azure web app to app service you don't lose the code associated, why should functions be different?

**Comment by balag0** [comment]
@onionhammer To startwith, does explicitly setting the WEBSITE_RUN_FROM_PACKAGE appsetting in the template during rerun fix this issue?
Also in the case of the web app template, how was the app deployed in the first place? Was it using Run From Package or some other mechanism?
(You are right we should expect to have the same behavior across function and web app, so trying to figure out what is different).

**Comment by onionhammer** [comment]
@balag0 Regarding your first question, yes, but I would have to know the (secret) URL to put in there at the time of deploying the ARM template.

In the case of web app deployment, we use a standard Zip Deploy (https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/deploy/azure-rm-web-app?view=azure-devops#deployment-methods), which is not supported for azure function consumption apps

**Comment by balag0** [comment]
oh I see, so in the case of standard zip deploy there are no explicit appsettings that configure the app contents, the contents are available on the remote disk directly. So not having the run from pkg appsetting explicitly as part of each deployment doesn't matter.
If you were to use RunFromPackage on the webapp, then I think you will run into this same limitation.

Can you explain a bit more on the secret URL? How is this url being generated? I assume you are not using RunFromPackage=1.
I know some skus dont support this (=1) yet but I think that could be the way forward to solving this issue.

**Comment by onionhammer** [comment]
@balag0 The difference is they're both the default/'auto' deployment mechanism for either platform:

For an app service webapp on Linux:

```yaml
  # Deploy via Azure DevOps
  - task: AzureWebApp@1
    inputs:
      azureSubscription: ${{ parameters.subscription }}
      appName: ${{ parameters.formWebApp }}
      appType: 'webAppLinux'
      slotName: ${{ parameters.slotName }}
      package: $(System.ArtifactsDirectory)/**/WebApp.zip
```

For an app service function app on Linux (consumption):

```yaml
  - task: AzureFunctionApp@1
    inputs:
      azureSubscription: ${{ parameters.subscription }}
      appType: 'functionAppLinux'
      appName: ${{ parameters.workersApp }}
      package:  $(System.ArtifactsDirectory)/**/C2C.Functions/*.zip
```

The only difference is your zip deployment will be lost if you re-run your ARM deployment on the function app but it wont be lost for the webapp deployment. What I mean by 'secret' is that the AzureFunctionApp deployment will stick a pre-authenticated azure blob storage URL to all your code into the 'WEBSITE_RUN_FROM_PACKAGE' config, so that config value cant be stored in your ARM template in source control.

**Comment by v-anvari** [comment]
cc @balag0

**Comment by devaraja666** [comment]
I was having the same issue, to work around it I'm using 2 bicep templates. on the first one i set up all the functions without WEBSITE_RUN_FROM_PACKAGE and extract all the functions keys and the second one updates the functions settings to include  WEBSITE_RUN_FROM_PACKAGE.

This was the only way i could found to get it working.
